Microsoft update 20180205

dhalbert opened the pull request:
Same pull request also sent to Microsoft (https://github.com/Microsoft/uf2-samd21/pull/31). This brings us up to date with v1.26.0 of Microsoft/uf2-samd21. Major changes there have been compilation improvements to save space and addition of WebUSB.

I tested this casually on a CPX and one other board (I forget which). I didn't test WebUSB specifically. This includes the latest upstream fixes that remove the WebUSB notification when you plug in the board.

tannewt commented on src/cdc_enumerate.c:
```diff
@@ -370,7 +366,7 @@ typedef struct {
 } StringDescriptor;
 
-static const char *string_descriptors[] = {0, VENDOR_NAME, PRODUCT_NAME, };
+static const char *string_descriptors[] = {0, VENDOR_NAME, PRODUCT_NAME, "4242"};
```
Shouldn't the serial number actually be unique?

dhalbert replied:
In theory, I guess, yes. That `"4242"` was added upstream. We could just leave it for now and open an issue, use a random number, or adapt your serial number code.

tannewt replied:
Ok, I'll hunt it down upstream and comment.

dhalbert replied:
I meant open our own issue. I'm not sure they care a lot, since it's only an issue if you're plugging in multiple boards at once, right? We could fix it eventually and contribute it to upstream.

tannewt replied:
Not necessarily. I'd worry about browsers and OSes storing settings based on it and applying them to other devices that are plugged in later.

ladyada commented:
reviewed, something amiss in the UF2/WebUSB handoff so not merging yet. @dhalbert is on the case :)

dhalbert commented:
Handover to bootloader from usermode was not working for WebUSB. Added `_binfo` to .ld scripts, and moved those uf2-specific .ld scripts to `scripts/`, along with the `*_self.ld` scripts.

dhalbert commented:
@tannewt Added serial number support:
```
Feb 10 15:56:01 salmonx kernel: [232859.546803] usb 1-1.4: new full-speed USB device number 99 using ehci-pci
Feb 10 15:56:01 salmonx kernel: [232859.658174] usb 1-1.4: New USB device found, idVendor=239a, idProduct=0018
Feb 10 15:56:01 salmonx kernel: [232859.658178] usb 1-1.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Feb 10 15:56:01 salmonx kernel: [232859.658180] usb 1-1.4: Product: CPlay Express
Feb 10 15:56:01 salmonx kernel: [232859.658181] usb 1-1.4: Manufacturer: Adafruit Industries
Feb 10 15:56:01 salmonx kernel: [232859.658183] usb 1-1.4: SerialNumber: 29A435A4504D5741372E3120FF190A16
```
Code adapted from CircuitPython serial numbe code, with a little tweaking to reduce code size Note that this one does not reverse the nibbles. I might change the CPy to do also generate it in nibble order.

dhalbert commented:
@ladyada WebUSB problems appear to be fixed now.

ladyada commented:
OK i checked and it works with WebUSB, loading circuitpy by hand, and arduino!

dhalbert commented:
@tannewt Feel free to review even if merged. See discussion in HipChat Sat around 5pm.
